File: src/chapters/matrix-factorization/content/code/cholesky.r
# Cholesky factorization A = L L^T of a symmetric positive-definite A.
cholesky <- function(A) {
  n <- nrow(A)
  L <- matrix(0, n, n)
  for (j in 1:n) {
    s <- if (j > 1) sum(L[j, 1:(j - 1)]^2) else 0
    L[j, j] <- sqrt(A[j, j] - s)
    if (j < n) {
      for (i in (j + 1):n) {
        s <- if (j > 1) sum(L[i, 1:(j - 1)] * L[j, 1:(j - 1)]) else 0
        L[i, j] <- (A[i, j] - s) / L[j, j]
      }
    }
  }
  L
}

A <- matrix(c(4, 2, -2, 2, 10, 2, -2, 2, 5), nrow = 3, byrow = TRUE)
L <- cholesky(A)
for (i in 1:nrow(L)) {
  print(round(L[i, ], 4))
}

File: src/chapters/matrix-factorization/content/code/cholesky.cpp
#include <vector>
#include <cmath>
#include <iostream>
using namespace std;
using Mat = vector<vector<double>>;

// Cholesky factorization A = L Lᵀ of a symmetric positive-definite A.
Mat cholesky(const Mat& A) {
    int n = A.size();
    Mat L(n, vector<double>(n, 0.0));
    for (int j = 0; j < n; ++j) {
        double s = A[j][j];
        for (int k = 0; k < j; ++k) s -= L[j][k] * L[j][k];
        L[j][j] = sqrt(s);
        for (int i = j + 1; i < n; ++i) {
            double t = A[i][j];
            for (int k = 0; k < j; ++k) t -= L[i][k] * L[j][k];
            L[i][j] = t / L[j][j];
        }
    }
    return L;
}

int main() {
    Mat A = {{4, 2, -2}, {2, 10, 2}, {-2, 2, 5}};
    for (auto& row : cholesky(A)) { for (double v : row) cout << v << " "; cout << "\n"; }
}
